assets/src/components/groupedAutocomplete.tsx

```diff
@@ -0,0 +1,731 @@
+    }
+
+    case CursorLocationAction.MoveToEnd: {
+      // Check if the last value of each length is valid
+      return getValidLocationOrUnsetWithLengthsContext({
+        group: lengthsContext.length - 1,
+        // move out of bounds ( < 0) if it doesn't exist
+        option: lengthsContext[lengthsContext.length - 1] - 1,
+      })
+    }
+  }
+
+  // Do actions that require the `cursorState`
+
+  // Return the current state if it's invalid
+  if (!cursorState) {
+    return cursorState
+  }
+
+  switch (action) {
+    case CursorLocationAction.FindNearestValidLocation: {
+      // Constrain current cursor to bounds of `lengthsContext`
+      return findNearestValidLocation()
+    }
+
+    case CursorLocationAction.MoveNext: {
+      const validCursorLocation =
+        getValidLocationOrUnsetWithLengthsContext({
+          ...cursorState,
+          option: cursorState.option + 1,
+        }) ||
+        // Then the first option in the next group
+        getValidLocationOrUnsetWithLengthsContext({
+          group: cursorState.group + 1,
+          option: 0,
+        })
+
+      if (validCursorLocation !== undefined) {
+        return validCursorLocation
+      }
+
+      // Otherwise return the current state as we've reached the end
+      onCursor?.onCursorExitEdge?.(CursorExitDirection.ExitEnd)
+      return cursorState
+    }
+
+    case CursorLocationAction.MovePrev: {
+      const validCursorLocation =
+        // Try the option index before the current option
+        getValidLocationOrUnsetWithLengthsContext({
+          ...cursorState,
+          option: cursorState.option - 1,
+        }) ||
+        // Try the last option of the previous group
+        getValidLocationOrUnsetWithLengthsContext({
+          group: cursorState.group - 1,
+          option: lengthsContext[cursorState.group - 1] - 1,
+        })
+
+      if (validCursorLocation) {
+        return validCursorLocation
+      }
+
+      onCursor?.onCursorExitEdge?.(CursorExitDirection.ExitStart)
+
+      return cursorState
+    }
+  }
+
+  // We really shouldn't reach this point, so for safety we'll return the
+  // current state
+  captureException(
+    new Error(
+      "internal error: entered unreachable code: end of `cursorLocationReducer`"
+    ),
+    {
+      extra: {
+        cursorState,
+        lengthsContext,
+        action,
+      },
+    }
+  )
+  return cursorState
+
+  /**
+   * Constrain {@link cursorState} onto the region defined by the
+   * {@link lengthsContext}.
+   */
+  function findNearestValidLocation() {
+    // Do not snap an undefined cursor.
+    if (cursorState === undefined) {
+      return undefined
+    }
+
+    // Clamp existing cursor into [0, Array.length) in each dimension
+    const group = clamp(cursorState.group, 0, lengthsContext.length - 1)
+    const option = clamp(cursorState.option, 0, lengthsContext[group] - 1)
+    // Check that answer is correct
```

Comment:
question: When could the answer not be correct?

Reply:
Uhhh, I don't have an answer for that! It's likely that it's always correct, I just didn't trust it at the time of writing.